docs/unidling.md

```diff
@@ -0,0 +1,188 @@
+3.  Resize the first R to *n* >= 1
+
+If no R is found no action is taken for the event to allow another controller to handle it.
+
+### Example: single kube-proxy receives single request for idled service
+
+This simplest possible unidle case is where an idled service S is unidled by a single request on a
+single node N.  
+
+1.  The kube-proxy on node N receives a request for S
+2.  The kube-proxy socket handler for S creates a `WantPods` event and waits X seconds for new
+    endpoints for S to be created
+3.  The unidler receives a watch event for the `WantPods` event
+4.  The unidler resolves the S to some replication controller R
+5.  The unidler resizes R to *n* >= 1
+6.  The replication controller manager schedules a new pod P for R
+7.  The scheduler determines placement and creates a `Binding` for P
+8.  The kubelet on the appropriate node processes the `Binding` and runs P
+9.  The endpoints controller determines that S has a new endpoint and updates S's endpoints
+10. The kube-proxy receives a watch event with the new endpoint information and updates its routing
+    table
+11. The kube-proxy services the request to the new endpoint
+
+### Thrash prevention
+
+The unidler must be able to correctly handle bursts of `WantPods` events for the same service.
+This implies that the unidler has an internal representation of which services have recently been
+unidled and that the unidle process itself has checks to ensure that only the minimum number of
+pods to unidle a service are created.  Concerns about scaling upon unidle should be handled
+exclusively by the autoscaler.
+
+### Example: multiple kube-proxies receive requests for idled service
+
+Thrash prevention becomes relevant when multiple kube-proxies receive requests for an idled service
+at approximately the same time.
+
+1.  The kube-proxy on node N1...Nx receive requests for S
+2.  The kube-proxy socket handlers create `WantPods` events and wait X seconds for new
+    endpoints for S to be created
+3.  The unidler receives the first watch event for the `WantPods` event
+4.  The unidler resolves the S to some replication controller R
+5.  The unidler resizes R to *n* >= 1
+6.  The unidler receives subsequent `WantPods` events from other nodes and ignores them since
+    there is already a replication controller that resolves to S with *n* >= 1
+7.  The replication controller manager schedules a new pod P for R
+8.  The scheduler determines placement and creates a `Binding` for P
+9.  The kubelet on the appropriate node processes the `Binding` and runs P
+10. The endpoints controller determines that S has a new endpoint and updates S's endpoints
+11. The kube-proxies receive watch events with the new endpoint information and updates their routing
+    tables
+12. The kube-proxies service the requests to the new endpoint
+
+## Proposed Design
+
+### Modifications to kube-proxy
+
+The kube-proxy must be modified create events when idled services are requested and make an attempt
```

Comment:
People will sometimes want to provide their own load-balancers in place of the kube-proxy, if the kube proxy does not provide enough features, or if they want to use something like https://cloud.google.com/compute/docs/load-balancing/.
So, kube-proxy should not be the only thing that can create "demand" events.